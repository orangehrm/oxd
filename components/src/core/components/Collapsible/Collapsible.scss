@import '../../../styles';
@import 'variables';

.oxd-collapsible-container {
  .oxd-collapsible-header-container {
    width: 100%;
    display: flex;
    justify-content: space-between;

    .oxd-collapsible-title {
      color: $oxd-collapsible-header-font-color;
      font-size: 16px;
      font-weight: 700;
      font-family: $oxd-collapsible-header-font-family;
      max-width: calc(100% - 24px);
      overflow-wrap: break-word;
      word-break: break-word;
    }

    .oxd-collapsible-caret-icon-wrapper {
      margin-right: 0;

      .oxd-collapsible-caret-icon {
        font-size: 5px;
        min-height: 24px;
        min-width: 24px;
      }
    }
  }
}
