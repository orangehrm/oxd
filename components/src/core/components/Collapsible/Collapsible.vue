<template>
  <div class="oxd-collapsible-container">
    <div class="oxd-collapsible-header-container">
      <label class="oxd-collapsible-title"> {{ $vt(headerTitle) }}</label>
      <div class="oxd-collapsible-caret-icon-wrapper">
        <oxd-icon-button
          :name="isExpanded ? 'oxd-caret-up' : 'oxd-caret-down'"
          @click="toggle"
          class="oxd-collapsible-caret-icon"
          size="xxx-small"
          :tooltip="isExpanded ? $vt('Show Less') : $vt('Show More')"
          v-if="collapsible"
        />
      </div>
    </div>
    <div class="oxd-collapsed-content" v-if="!isExpanded">
      <slot name="collapsedContent"></slot>
    </div>
    <div class="oxd-expanded-content" v-if="isExpanded">
      <slot name="expandedContent"></slot>
    </div>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import oxdIconButton from '@orangehrm/oxd/core/components/Button/Icon.vue';
import translateMixin from '../../../mixins/translate';

export default defineComponent({
  name: 'CollapsibleComponent',
  components: {
    'oxd-icon-button': oxdIconButton,
  },
  mixins: [translateMixin],
  props: {
    headerTitle: {
      type: String,
      default: '',
    },
    defaultExpand: {
      type: Boolean,
      default: false,
    },
    collapsible: {
      type: Boolean,
      default: true,
    },
  },
  data() {
    return {
      isExpanded: this.defaultExpand,
    };
  },
  methods: {
    toggle() {
      this.isExpanded = !this.isExpanded;
      this.$emit('toggle', {isExpanded: this.isExpanded});
    },
  },
});
</script>

<style src="./Collapsible.scss" lang="scss" scoped></style>
