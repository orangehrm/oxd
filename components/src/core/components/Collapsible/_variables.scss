$oxd-collapsible-header-font-color: $oxd-interface-gray-darken-1-color !default;
$oxd-collapsible-header-font-family: $oxd-font-family !default;
