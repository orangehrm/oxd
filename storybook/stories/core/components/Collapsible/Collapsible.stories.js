import Collapsible from '@orangehrm/oxd/core/components/Collapsible/Collapsible.vue';
import CollapsibleEvents from './CollapsibleEvents.story.vue';

export default {
  title: 'Layout/Collapsible',
  component: Collapsible,
  argTypes: {
    headerTitle: {
      control: {type: 'text'},
      table: {
        type: {summary: 'This is the title label of the collapsible'},
      },
    },
    defaultExpand: {
      control: {type: 'boolean'},
      table: {
        type: {
          summary:
            'This is to define whether the content should be expanded or collapsed initially',
        },
      },
    },
    collapsible: {
      control: {type: 'boolean'},
      table: {
        type: {
          summary:
            'This is to define whether the collapse/expand button should be shown or not',
        },
      },
    },
    toggle: {
      control: {type: 'object'},
      defaultValue: {},
      table: {
        type: {
          summary:
            'Emit event when click on the toggle button. Object contains the state whether the content is expanded or not',
        },
      },
    },
    collapsedContent: {
      control: {type: 'object'},
      table: {
        type: {summary: 'The slot which defines the collapsed content'},
      },
    },
    expandedContent: {
      control: {type: 'object'},
      table: {
        type: {summary: 'The slot which defines the expanded content'},
      },
    },
  },
};

const Template = (args) => ({
  setup() {
    return {args};
  },
  components: {'oxd-collapsible': Collapsible},
  template:
    '<oxd-collapsible v-bind="args"> <template #collapsedContent> <div> This is the collapsed view </div> </template> <template #expandedContent> <div> This is the expanded view </div> </template>  </oxd-collapsible>',
});

export const Default = Template.bind({});
Default.args = {
  headerTitle: 'Collapsible Component',
};

Default.parameters = {
  docs: {
    source: {
      code: '<oxd-collapsible \n :headerTitle="Collapsible Component"' + '\n/>',
    },
  },
};

export const InitialExpandedView = Template.bind({});
InitialExpandedView.args = {
  headerTitle: 'Collapsible Component',
  defaultExpand: true,
};

InitialExpandedView.parameters = {
  docs: {
    source: {
      code:
        '<oxd-collapsible \n :headerTitle="Collapsible Component"' +
        '\n/> \n :defaultExpand=true',
    },
  },
};

export const NoButtonToExpandOrCollapse = Template.bind({});
NoButtonToExpandOrCollapse.args = {
  headerTitle: 'Collapsible Component',
  collapsible: false,
};

NoButtonToExpandOrCollapse.parameters = {
  docs: {
    source: {
      code:
        '<oxd-collapsible \n :headerTitle="Collapsible Component"' +
        '\n/> \n :collapsible=false',
    },
  },
};

export const Events = () => CollapsibleEvents;

Events.parameters = {
  docs: {
    source: {
      code:
        '  <div> \n' +
        '<oxd-collapsible \n' +
        ':headerTitle="Collapsible Component"\n' +
        '@toggle="onToggle""\n' +
        '>\n' +
        '<template #collapsedContent>\n' +
        '<div>This the Collapsed View</div>\n' +
        '</template>\n' +
        '<template #expandedContent>\n' +
        '<div>This the Expanded View</div>\n' +
        '</template>\n' +
        '</oxd-collapsible>\n' +
        '</div>\n' +
        '<div style="margin-top: 2rem">\n' +
        '<span v-if="ToggleEvent">Toggle Event Triggered <br> Toggle data {{ToggleInfo}}</span>' +
        '</div>\n' +
        '//\n' +
        'File -> CollapsibleEvents.story.vue',
    },
  },
};
