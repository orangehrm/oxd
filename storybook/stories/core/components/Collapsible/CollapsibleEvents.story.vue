<template>
  <div>
    <oxd-collapsible
      :headerTitle="'Collapsible Component'"
      @toggle="onToggle"
    >
    <template #collapsedContent>
      <div>This the Collapsed View</div>
    </template>
    <template #expandedContent>
      <div>This the Expanded View</div>
    </template>
    </oxd-collapsible>
  </div>
  <div style="margin-top: 2rem">
    <span v-if="ToggleEvent">Toggle Event Triggered <br> Toggle data {{ToggleInfo}}</span>
  </div>
</template>

<script>
import Collapsible from '@orangehrm/oxd/core/components/Collapsible/Collapsible.vue';
import { ref } from "vue";

export default {
  data() {
    return {
      ToggleEvent: false,
      ToggleInfo : "",
    };
  },

  components: {
    'oxd-collapsible': Collapsible,
  },
  methods: {
    clearFields() {
      this.ToggleEvent = false;

    },
    onToggle(data) {
      this.clearFields();
      this.ToggleEvent = true;
      this.ToggleInfo=data;
    },
  },
};
</script>
