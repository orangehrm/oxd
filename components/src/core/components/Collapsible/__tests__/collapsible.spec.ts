import {shallowMount} from '@vue/test-utils';
import Collapsible from '@orangehrm/oxd/core/components/Collapsible/Collapsible.vue';

describe('Collapsible', () => {
  it('renders the collapsible component', () => {
    const wrapper = shallowMount(Collapsible, {});
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('header title label', () => {
    const wrapper = shallowMount(Collapsible, {
      props: {
        headerTitle: 'Candidate History',
      },
    });
    expect(wrapper.html()).toContain('Candidate History');
  });

  it('when collapsed the collapsed slot is shown and the expanded slot is not shown', () => {
    const wrapper = shallowMount(Collapsible, {
      slots: {
        collapsedContent: '<div>This is the collapsed view</div>',
        expandedContent: '<div>This is the expanded view</div>',
      },
    });
    expect(wrapper.find('.oxd-collapsed-content').exists()).toBeTruthy();
    expect(wrapper.find('.oxd-expanded-content').exists()).toBeFalsy();
  });

  it('when expanded the expanded slot is shown and the collapsed slot is not shown', () => {
    const wrapper = shallowMount(Collapsible, {
      props: {
        defaultExpand: true,
      },
      slots: {
        collapsedContent: '<div>This is the collapsed view</div>',
        expandedContent: '<div>This is the expanded view</div>',
      },
    });
    expect(wrapper.find('.oxd-expanded-content').exists()).toBeTruthy();
    expect(wrapper.find('.oxd-collapsed-content').exists()).toBeFalsy();
  });

  it('content is expanded when click on carrot down button', async () => {
    const wrapper = shallowMount(Collapsible, {});
    const button = wrapper.find('.oxd-collapsible-caret-icon');
    await button.trigger('click');
    expect(wrapper.vm.isExpanded).toBe(true);
    expect(wrapper.find('.oxd-expanded-content').exists()).toBeTruthy();
  });

  it('content is collapsed when click on carrot up button', async () => {
    const wrapper = shallowMount(Collapsible, {
      props: {
        defaultExpand: true,
      },
    });
    const button = wrapper.find('.oxd-collapsible-caret-icon');
    await button.trigger('click');
    expect(wrapper.vm.isExpanded).toBe(false);
    expect(wrapper.find('.oxd-collapsed-content').exists()).toBeTruthy();
  });

  it('no collapse/expand button is there', () => {
    const wrapper = shallowMount(Collapsible, {
      props: {
        collapsible: false,
      },
    });
    expect(wrapper.find('.oxd-collapsible-caret-icon').exists()).toBeFalsy();
  });
});
