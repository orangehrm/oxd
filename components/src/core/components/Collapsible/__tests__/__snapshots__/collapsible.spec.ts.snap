// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapsible renders the collapsible component 1`] = `
<div class="oxd-collapsible-container">
  <div class="oxd-collapsible-header-container"><label class="oxd-collapsible-title"></label>
    <div class="oxd-collapsible-caret-icon-wrapper">
      <oxd-icon-button-stub name="oxd-caret-down" size="xxx-small" withcontainer="true" disabled="false" tooltip="Show More" flow="top" iconstyles="[object Object]" showtooltipwhendisabled="false" class="oxd-collapsible-caret-icon"></oxd-icon-button-stub>
    </div>
  </div>
  <div class="oxd-collapsed-content"></div>
  <!--v-if-->
</div>
`;
